### ISSUE TITLE
[Feature] "yarn workspaces focus" option `--prune-cache`

### ISSUE THREAD

**Issue by yinzara** [yarnpkg/berry#1789]
- [X ] I'd be willing to implement this feature (maybe) 👍 
- [X ] This feature can already be implemented through a plugin modification

**Describe the user story**

In my automated build environment for a monorepo with around 40 packages, when I'm packaging services in Docker, with Yarn Classic I used to use the `yarn install --production` command to install only the dependencies necessary to run my service.

The documentation for yarn berry says to use the `yarn workspaces focus --production` command as a replacement for `yarn install --production`.  Unfortunately this isn't helpful because the ".yarn/cache" is not pruned (any devDependencies still remain).  This means my Docker images now contain the zip files for all the devDependencies when they should not increasing their size by multiples.  

My only option is to not copy the .yarn/cache and .yarn/unplugged directories into my Docker image and rebuild them inside Docker.  The completely negates the benefits of already having all the dependencies available.

**Describe the solution you'd like**

I propose adding a new "--prune-cache" option.  This would forcibly prune ".yarn/cache" of any dependencies that didn't fit the other passed parameters (i.e. if no "-all" parameter is passed, only dependencies (and transitive) of the current project will be in ".yarn/cache", if "--production" is passed, only "dependencies" and not "devDependencies" will be included).  

I would run this right before I ran my Docker build so that when I copied in my .yarn directory, it would include all my dependencies and I wouldn't have to re-download them.

**Describe the drawbacks of your solution**

Adding another flag adds future maintenance and complexity.  Using this flag actually deletes files in your current cache meaning your local workspace will have to download them again if they change focus.

**Describe alternatives you've considered**

Instead of copying the entire ".yarn" directory, only copy the .yarn/release and .yarn/plugins into the Docker image then run `yarn workspaces focus --production` in the Docker image which downloads only the "dependencies".  This means I now have to download all my dependencies again from my registry.

**Comment by yinzara** [comment]
Looking at the code directly this shouldn't be too hard to implement but would require a change to the "InstallOptions" interface in berry/packages/yarnpkg-core/sources/Project.ts to have a new "cleanCache" or "pruneCache" option that would only be used when "persistProject" was false that would cause "await this.cacheCleanup()" to be called but not "await this.persist()".

Then modify the plugin-workspace-tools to have "focus" pass the "cleanCache" option based on the "--prune-cache" flag

**Comment by arcanis** [comment]
Could be interesting - however I've also heard people who wanted to wanted to remove the files from the workspaces that didn't get selected, so that the `COPY` operation would properly compute the cache key for the layer (whereas it would otherwise change each time any workspace is modified).

In this context, I wonder if perhaps we shouldn't do a single `--prune-project` flag that would completely remove all files that aren't part of the selected workspaces. Do you have thoughts on that?

**Comment by yinzara** [comment]
Well that's an interesting addition, but I think might need to be a separate feature as it would require a pretty substantial change to the core I think (at least that's what I'd guess, I'm not 100% sure).   Right now yarn doesn't modify any files outside of .yarn except for the project.json and yarn.lock files (at least that I know of).  I also think it's a slightly different use case, though still related.  Since workspaces are directories, you can already KINDA do this by only selectively copying directories.  You can't do it for dependencies though if you want to use --production.

My suggestion can be done with only a few lines of changes to the workspace plugin.

**Comment by yinzara** [comment]
Honestly that change IMO is also super easy. If we just add an "immutable" flag as you suggest and pass it to the "install", it should do exactly as you say.  @arcanis does that warrant a separate PR?

**Comment by merceyz** [comment]
> does that warrant a separate PR?

I'd say so, as it's a separate feature

**Comment by yinzara** [comment]
Yes and no. They're both on the same component and both relate to the same use case, Docker installs.  So there is an argument that they could be included in the same PR.  If I changed this issue and PR to be "Optimize `yarn workspaces focus` for Docker installations" then it could all fall under the same issue.

**Comment by arcanis** [comment]
Please open a different issue for this. Features tend to have costs higher than just the raw implementation costs, and in this particular case I have reserves that should be discussed separately.

**Comment by andreialecu** [comment]
> Could be interesting - however I've also heard people who wanted to wanted to remove the files from the workspaces that didn't get selected, so that the COPY operation would properly compute the cache key for the layer (whereas it would otherwise change each time any workspace is modified).

> In this context, I wonder if perhaps we shouldn't do a single --prune-project flag that would completely remove all files that aren't part of the selected workspaces. Do you have thoughts on that?

I ended up asking about this same feature in the Discord chat earlier today. 

My suggestion was to implement some command that creates a copy of the project, minus the unrelated workspaces (and their dependencies). Either to a configurable output directory or to `/tmp/`. From my understanding, this can be implemented as a separate plugin, and once stable enough, it can be probably merged back to core.

This way there's no risk of accidental data loss. You'd have to specify a separate (empty) directory to focus to.

There is some "prior art" here, `yarn patch` creates a `/tmp/` folder with a copy of a dependency.

**Comment by Larry1123** [comment]
While this issue's topic is about adding a flag to prune the current install. The reason for wanting this feature is quite why I had created a plugin for creating minimal production installs for use in docker. It creates a new yarn cache with hard links to the packages and will pack any `workspace:` dependency into this new cache. I am still working on adding features to it but it is stable and I have been using in builds. https://gitlab.com/Larry1123/yarn-contrib/-/tree/master/packages/plugin-production-install If something like this is better for your use case but it does not meet everything feel free to open an issue on it's repo or reach out on discord.

**Comment by asvetliakov** [comment]
@Larry1123 this is exactly what i was needed, thank you! This should be built into yarn itself, nobody wants to copy 300 MB of dev packages into production images or lambda

**Comment by Larry1123** [comment]
Just an fyi, with how docker works removing files in a later layer does not remove them from the docker image overall. The resulting file system will not have them yes. However it is still part of the image. Same thing happens if you `chown` files it adds them a second time, and doubles (of the files touched) the resulting total size of the image. The only way around it is if you using multipart builds, and copy only the files you want on top of a new slim final base image. I don't think that the method you want to take truly solves your issue.
The reason my plugin creates a new output is so that can be copied into a clean layer without worrying about lingering layers.

**Comment by jtbennett** [comment]
@Larry1123 That's a good point. I typically use a multi-stage Dockerfile. All the work happens in a build stage, then the final image is just copying files from that build. It keeps the final image as small as possible.

This Dockerfile is a template and not necessarily usable as-is, but it demonstrates the approach:
https://github.com/jtbennett/create-ts-project/blob/main/packages/create-ts-project/template/docker/Dockerfile

**Comment by yinzara** [comment]
@Larry1123 oh don't worry I'm very aware of that fact.  If you happen to have a build automation build system that runs via Dockerfile (like pretty much all of them now), you can have your build image based on the same OS and then you run `yarn workspaces focus....` before you even start the Docker build process and then copy the .yarn directory into the Docker image with the pruned cached.

**Comment by arcanis** [comment]
Closing as wontfix due to https://github.com/yarnpkg/berry/pull/1798#issuecomment-713030607

**Comment by yinzara** [comment]
Well I can't say I'm happy. More functionality that's available in Yarn 1 (and honestly a very important part of it IMHO) that isn't available in Yarn 2.0 without custom plugins by third parties. Great.  By the time I convert my project from Yarn 1 to Yarn 2 it's all just going to be custom code.

**Comment by arcanis** [comment]
I'm sorry about this. We're all volunteers, and we need to be mindful how we spend our resources (including maintenance time). If we don't intend to support something, what's the point of having it in the core, especially if it can live elsewhere just fine? At least this way its development isn't tied to our release schedule.

> I can't see why you would encourage people to use a Zero-Installs setup and then not provide a way to clean up dev dependencies.

Your request completely relies on 1/ a docker environment 2/ an unwillingness to use the plugin that @Larry1123 is building. It's not unreasonable, but you have to understand it isn't the typical situation most of our users are in, and as such it doesn't necessarily make sense to add a feature just for you.

Yarn 1 precisely had this problem that, since it wasn't modular, we *had* to merge features into the core even if they only mattered to few people, otherwise they wouldn't be able to move forward. As a result, over time, Yarn 1 became bloated, complex, with many edge cases, and maintenance became a nightmare. I don't want that to happen again, hence why we designed the v2 with a core suited to receive features from the outside. If you want something but we're not convinced, you don't have to be blocked by us anymore, but the reverse is true: we don't have to merge everything.

**Comment by yinzara** [comment]
< Your request completely relies on 1/ a docker environment 2/ an unwillingness to use the plugin that @Larry1123 is building

Docker is only a single example of this use case but there are a myriad. Anytime you have a system where the file system storage space matters (i.e. anytime you're using any kind of a FS image) ensuring there is a method to minimize the use of said filesystem is important. 

Larry's solution is an extremely specific in-depth choice as to a method for packaging Docker images in a workspace.  I argue that if he had 100 microservices and 10 shared libraries using this method, he would have enormous amount of waste of nearly identical layers because he's copying the dependencies into the image for each microservice and that a completely different approach of maintaining consistent dependency versions across the workspace and a shared base image containing all the dependencies would be a much more efficient approach both in the build time and in the final image sizes.

So as you are implying, there might be many ways of solving the problem but saying I'm simply refusing to use @Larry1123 solution because it's "the solution for Docker" is incorrect.

My request is abstract and very simple.

Provide a way with Zero-Installs to remove "devDependencies" in the files stored on disk.  That's all. Even the package.json considers these things so important as to separate them out.  Every other package manager for JS does it (yarn 1, npm, npnm).

`yarn install --production` is probably one of the most used yarn commands in every automated build system that uses yarn.  There is no equivalent command in yarn 2 if you use Zero-Installs.  Why is it when I use the "preferred" configuration of Yarn 2, there is no way to achieve that without using third party tools?

**Comment by andreialecu** [comment]
@yinzara the approach of v1 and npm as far as production installs go wasn't really ideal, but there was no easy alternative because `node_modules` was huge. 

With the PnP installation mechanism and zip files, it's finally feasible to run a production install into a separate directory.

Your sample above is much more easily achievable using the plugin made by @Larry1123 [plugin-production-install](https://gitlab.com/Larry1123/yarn-contrib/-/tree/master/packages/plugin-production-install)

And it does a lot of other cool things, like remove unneeded extra files that yarn keeps around, and it packs other workspaces for you into their own zip file.

```ts
yarn
yarn build
yarn test

yarn prod-install --pack /tmp/build
s3 sync /tmp/build s3://my-bucket
```

**Comment by arcanis** [comment]
@yinzara It's late at night, I already took a lot of time to write you detailed answers and I honestly don't know how to be clearer: **if you feel strongly about a feature, prove us wrong by implementing it on your end, maintaining it, and showing that it solves a use case better than other options**. Don't try to sell it repeatedly to someone that wasn't convinced the first three times you explained, it doesn't help anyone.

Docker is a **recurrent** problem, it existed back in Yarn 1 already, and I strongly suspect it's not as simple as "just have --prune-cache". If it is for you that's fine, but, again: **put it in practice and prove it on your end**. In the meantime I'll have to lock this thread, things are going circle.
